clippy_lints/src/loops.rs

```diff
@@ -485,140 +491,107 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Pass {
+        ExprIndex(ref e1, ref e2) => never_loop_expr_seq(&mut [&**e1, &**e2].iter().cloned(), state),
         ExprIf(ref e, ref e2, ref e3) => {
-            [e, e2].iter().chain(e3.as_ref().iter()).any(|e| {
-                contains_continue_expr(e, dest)
-            })
+            let e1 = never_loop_expr(e, state);
+            let e2 = never_loop_expr(e2, state);
+            match *e3 {
+                Some(ref e3) => {
+                    let e3 = never_loop_expr(e3, state);
+                    e1 && (e2 || e3)
+                },
+                None => e1,
+            }
+        },
+        ExprLoop(ref b, _, _) => {
+            let block_may_complete = never_loop_block(b, state);
+            let has_break = state.breaks.remove(&expr.id);
+            state.continues.remove(&expr.id);
+            block_may_complete || has_break
+        },
+        ExprWhile(ref e, ref b, _) => {
+            let e = never_loop_expr(e, state);
+            let block_may_complete = never_loop_block(b, state);
+            let has_break = state.breaks.remove(&expr.id);
+            let has_continue = state.continues.remove(&expr.id);
+            e && (block_may_complete || has_break || has_continue)
         },
-        ExprWhile(ref e, ref b, _) => contains_continue_expr(e, dest) || contains_continue_block(b, dest),
         ExprMatch(ref e, ref arms, _) => {
-            contains_continue_expr(e, dest) || arms.iter().any(|a| contains_continue_expr(&a.body, dest))
+            let e = never_loop_expr(e, state);
+            let arms = never_loop_expr_branch(&mut arms.iter().map(|a| &*a.body), state);
+            e && arms
         },
-        ExprBlock(ref block) |
-        ExprLoop(ref block, ..) => contains_continue_block(block, dest),
-        ExprStruct(_, _, ref base) => {
-            base.as_ref().map_or(
-                false,
-                |e| contains_continue_expr(e, dest),
-            )
+        ExprBlock(ref b) => never_loop_block(b, state),
+        ExprAgain(d) => {
+            let id = d.target_id.opt_id().expect("continue is missing target id");
+            state.continues.insert(id);
+            false
         },
-        ExprAgain(d) => dest.map_or(true, |dest| d.target_id.opt_id().map_or(false, |id| id == dest)),
-        _ => false,
-    }
-}
-
-fn loop_exit_block(block: &Block, loops: &mut Vec<NodeId>) -> bool {
-    block.stmts.iter().take_while(|s| !contains_continue_stmt(s, None)).any(|s| loop_exit_stmt(s, loops))
-        || block.expr.as_ref().map_or(false, |e| loop_exit_expr(e, loops))
-}
-
-fn loop_exit_stmt(stmt: &Stmt, loops: &mut Vec<NodeId>) -> bool {
-    match stmt.node {
-        StmtSemi(ref e, _) |
-        StmtExpr(ref e, _) => loop_exit_expr(e, loops),
-        StmtDecl(ref d, _) => loop_exit_decl(d, loops),
+        ExprBreak(d, _) => {
+            let id = d.target_id.opt_id().expect("break is missing target id");
```

Comment:
is this okay?

Reply:
I'm not sure. my gut says no, but Travis passes... I'd need to look at some code or into rustc, but I'm on mobile right now.

Reply:
Ok, I checked, yes this is fine, the only way it can fail is if there are errors (like `while continue {}` or `while break {}`)

Change the expect message to "target id can only be missing in the presence of compilation errors".

Reply:
Great! Done.